File: src/gridcell.cpp
#include "gridcell.h"

GridCell::GridCell() {
	this->mAreaType = Area_Type::Nothing;
	this->mVerticalWallType = Wall_Type::Nothing;
	this->mHorizontalWallType = Wall_Type::Nothing;
	this->mPointType = Point_Type::Nothing;
}

GridCell::GridCell(const Area_Type area) {
	this->mAreaType = area;
	this->mVerticalWallType = Wall_Type::Nothing;
	this->mHorizontalWallType = Wall_Type::Nothing;
	this->mPointType = Point_Type::Nothing;
}

GridCell::GridCell(const Area_Type area, const Wall_Type horizontal_wall, const Wall_Type vert_wall) {
	this->mAreaType = area;
	this->mVerticalWallType = vert_wall;
	this->mHorizontalWallType = horizontal_wall;
	this->mPointType = Point_Type::Nothing;
}

GridCell::GridCell(const Area_Type area, const Wall_Type horizontal_wall, const Wall_Type vert_wall, const Point_Type point) {
	this->mAreaType = area;
	this->mVerticalWallType = vert_wall;
	this->mHorizontalWallType = horizontal_wall;
	this->mPointType = point;
}

void GridCell::setAccessable(Accessable_t acc)
{
	this->mAccessable = acc;
}

void GridCell::setArea(const Area_Type area) {
	this->mAreaType = area;
}

void GridCell::setVertWall(const Wall_Type wall) {
	this->mVerticalWallType = wall;
}

void GridCell::setHorizWall(const Wall_Type wall) {
	this->mHorizontalWallType = wall;
}

void GridCell::setPoint(const Point_Type point) {
	this->mPointType = point;
}

bool GridCell::isAccessable()
{
	if (this->mAccessable == Accessable_t::Yes) {
		return true;
	} else {
		return false;
	}
}

bool GridCell::isRoom() {
	if (this->mAreaType == Area_Type::Room) {
		return true;
	} else {
		return false;
	}
}

bool GridCell::isEntrance() {
	if (this->mAreaType == Area_Type::Entrance) {
		return true;
	} else {
		return false;
	}
}

bool GridCell::isEmpty() {
	if (this->mAreaType == Area_Type::Nothing && 
			this->mVerticalWallType == Wall_Type::Nothing &&
			this->mHorizontalWallType == Wall_Type::Nothing &&
			this->mPointType == Point_Type::Nothing) {
		return true;
	} else {
		return false;
	}
}

Area_Type GridCell::getArea() {
	return this->mAreaType;
}

Wall_Type GridCell::getVertWall() {
	return this->mVerticalWallType;
}

Wall_Type GridCell::getHorizWall() {
	return this->mHorizontalWallType;
}

Point_Type GridCell::getPoint() {
	return this->mPointType;
}

File: src/rpgtools/map/gridmap/gridcell.rs
#[derive(Clone, Debug, PartialEq)]
pub enum AreaType {
    Nothing,
    Entrance,
    Room,
    //    Stairs,
    //    Tested,
}

#[derive(Clone, Debug)]
pub enum WallType {
    Nothing,
    //    Wall,
    //    Door,
    //    SecretDoor,
}

#[derive(Clone, Debug)]
pub enum PointType {
    Nothing,
    //    Pillar,
}

/// Representation of a GridCell, which is a single unit in a grid.
#[derive(Clone, Debug)]
pub struct GridCell {
    /// The type of area contained within this sell
    pub area: AreaType,
    vert_wall: WallType,
    horiz_wall: WallType,
    point: PointType,
}

impl GridCell {
    /// Construct a new GridCell
    pub fn new() -> GridCell {
        GridCell {
            area: AreaType::Nothing,
            vert_wall: WallType::Nothing,
            horiz_wall: WallType::Nothing,
            point: PointType::Nothing,
        }
    }

    /// Check if a GridCell is 'empty', meaning that all drawing-related
    /// fields have a value of 'Nothing'.
    pub fn is_empty(&self) -> bool {
        matches!(
            *self,
            GridCell {
                area: AreaType::Nothing,
                vert_wall: WallType::Nothing,
                horiz_wall: WallType::Nothing,
                point: PointType::Nothing,
                ..
            }
        )
    }

    //    /// Check whether the GridCell represents a dungeon entrance
    //    pub fn is_entrance(&self) -> bool {
    //        match self {
    //            &GridCell { area: AreaType::Entrance, ..} => true,
    //            _ => false
    //        }
    //    }

    /// Check whether the GridCell is part of a dungeon room
    pub fn is_room(&self) -> bool {
        !matches!(
            *self,
            GridCell {
                area: AreaType::Nothing,
                ..
            }
        )
    }

    /// Get the type of area that this cell represents
    pub fn area(&self) -> &AreaType {
        &self.area
    }
}

impl Default for GridCell {
    fn default() -> Self {
        Self::new()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn new() {
        let cell = GridCell::new();
        assert_eq!(AreaType::Nothing, cell.area);
    }

    #[test]
    fn is_empty() {
        let mut cell = GridCell::new();
        // Do checks
        cell.area = AreaType::Nothing;
        assert_eq!(true, cell.is_empty());
        cell.area = AreaType::Entrance;
        assert_eq!(false, cell.is_empty());
        cell.area = AreaType::Room;
        assert_eq!(false, cell.is_empty());
    }

    #[test]
    fn is_room() {
        let mut cell = GridCell::new();
        // Do checks
        cell.area = AreaType::Nothing;
        assert_eq!(false, cell.is_room());
        cell.area = AreaType::Entrance;
        assert_eq!(true, cell.is_room());
        cell.area = AreaType::Room;
        assert_eq!(true, cell.is_room());
    }
}
